File: streampark-flink/streampark-flink-connector/streampark-flink-connector-jdbc/src/main/scala/org/apache/streampark/flink/connector/jdbc/request/JdbcRequest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.jdbc.request

import org.apache.streampark.common.util.Utils
import org.apache.streampark.flink.connector.jdbc.internal.JdbcASyncFunction
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.api.scala.{AsyncDataStream, DataStream}

import java.util.Properties
import java.util.concurrent.TimeUnit
import scala.annotation.meta.param

object JdbcRequest {

  def apply[T: TypeInformation](@(transient@param) stream: DataStream[T], property: Properties = new Properties()): JdbcRequest[T] = new JdbcRequest[T](stream, property)

}

class JdbcRequest[T: TypeInformation](@(transient@param) private val stream: DataStream[T], property: Properties = new Properties()) {

  /**
   *
   * @param sqlFun
   * @param jdbc
   * @tparam R
   * @return
   */
  def requestOrdered[R: TypeInformation](@(transient@param) sqlFun: T => String, @(transient@param) resultFun: (T, Map[String, _]) => R, timeout: Long = 1000, capacity: Int = 10)(implicit jdbc: Properties): DataStream[R] = {
    Utils.copyProperties(property, jdbc)
    val async = new JdbcASyncFunction[T, R](sqlFun, resultFun, jdbc)
    AsyncDataStream.orderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity)
  }

  def requestUnordered[R: TypeInformation](@(transient@param) sqlFun: T => String, @(transient@param) resultFun: (T, Map[String, _]) => R, timeout: Long = 1000, capacity: Int = 10)(implicit jdbc: Properties): DataStream[R] = {
    Utils.copyProperties(property, jdbc)
    val async = new JdbcASyncFunction[T, R](sqlFun, resultFun, jdbc)
    AsyncDataStream.unorderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity)
  }

}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-jdbc/src/main/java/org/apache/streampark/flink/connector/jdbc/request/JdbcRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.jdbc.request;

import org.apache.streampark.common.util.Utils;
import org.apache.streampark.flink.connector.function.TransformFunction;
import org.apache.streampark.flink.connector.jdbc.internal.JdbcASyncFunction;

import org.apache.flink.streaming.api.datastream.AsyncDataStream;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;

import java.util.Properties;
import java.util.concurrent.TimeUnit;

/** JDBC async request helper. */
public class JdbcRequest<T> {

    private final DataStream<T> stream;
    private final Properties property;

    public JdbcRequest(DataStream<T> stream, Properties property) {
        this.stream = stream;
        this.property = property != null ? property : new Properties();
    }

    public static <T> JdbcRequest<T> of(DataStream<T> stream, Properties property) {
        return new JdbcRequest<>(stream, property);
    }

    public <R> SingleOutputStreamOperator<R> requestOrdered(
            TransformFunction<T, String> sqlFun,
            JdbcASyncFunction.JdbcResultFunction<T, R> resultFun,
            long timeout,
            int capacity,
            Properties jdbc) {
        Utils.copyProperties(property, jdbc);
        JdbcASyncFunction<T, R> async = new JdbcASyncFunction<>(sqlFun, resultFun, jdbc, capacity);
        return AsyncDataStream.orderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity);
    }

    public <R> SingleOutputStreamOperator<R> requestUnordered(
            TransformFunction<T, String> sqlFun,
            JdbcASyncFunction.JdbcResultFunction<T, R> resultFun,
            long timeout,
            int capacity,
            Properties jdbc) {
        Utils.copyProperties(property, jdbc);
        JdbcASyncFunction<T, R> async = new JdbcASyncFunction<>(sqlFun, resultFun, jdbc, capacity);
        return AsyncDataStream.unorderedWait(stream, async, timeout, TimeUnit.MILLISECONDS, capacity);
    }
}
